### PR TITLE
New capsule@.service feature

### PR DESCRIPTION
**Pull request by poettering**
People are using systemd --user to run applications unprivileged. Let's make this official.

This adds a new unit systemd-project@.service, which is very similar to user@.service. However it is also a bit different:

* It sets DynamicUser=1 to generate a dynamic user while it runs
* It is not tracked by logind
* It runs without PAM session
* It has HOME set to /var/lib/projects/%i and XDG_RUNTIME_DIR set to /run/projects/%i

With this in place it's sufficient to do "systemctl start systemd-project@foobar.service" to get a new "project" called "foobar" running, with its own scratch dir under /var/lib/projects/

The suggested mode of operation is that people set up services in `/var/lib/projects/<project>/.config/systemd/user/`, and then start the thing via the "systemctl start".

I think the concept makes a ton of sense, and is very very simple.

This also adds support for a new --project=/-J switch to systemctl and systemd-run to talk to these "projects" from the outside.

This is an RFC at the moment, because there are a couple of question to answer:

1. Do we actually want this? It might confuse people, since they now have have another way to bundle up things:
  a. As regular services (possibly via confext/sysext)
  b. As portable services
  c. As nspawn containers
  d. As projects
2. dbus config prohibits access to all foreign users, including those with UID 0. Which means you cannot actually talk to the dbus instance we start for the project from the outside. I haven't found a nice way to relax this policy for project busses without also relaxing it for user@.service, i.e. regular users.
3. I am not even sure about the name of the concept, is "project" really that great?
4. Which slice should projects live in by default? in system.slice? in user.slice? in a new top-level slice? For now I picked system-projects.slice, but I am not sure this is the best approach.
5. Should we maybe default to start something other than `default.target` in the service manager? maybe `project@%i.target`? After all, this service manager is mostly just another --user instance, hence if people add services to system-wide per-user default.target they might be quite confused what they get.
6. Should the service maybe be called `project@.service` rather than `systemd-project@.service`? The rules around naming services are a bit blurry, but the idea was always that generic concepts should not be prefixed with `systemd-`, and systemd specific implementations should be. From logind's perspective user@.service was generic and hence has no prefix. Question is what to do about the project service manager? with or without prefix?

ideas for all these questions more than welcome!

In general I was surprised how trivial this was to put together, the unit file I wrote for this 
just worked immediately.

### CODE REVIEW

File: src/shared/bus-util.c [hunk 1]
```diff
@@ -267,6 +267,60 @@ int bus_connect_user_systemd(sd_bus **ret_bus) {
...
+
+        *ret_bus = TAKE_PTR(bus);
+        return 0;
+}
+
+int bus_connect_project_bus(const char *project, sd_bus **ret_bus) {
+        _cleanup_(sd_bus_close_unrefp) sd_bus *bus = NULL;
+        _cleanup_free_ char *pp = NULL;
+        int r;
+
```
Comment (bluca) [inline src/shared/bus-util.c:355]: assert(ret_bus)

File: src/shared/bus-util.c [hunk 2]
```diff
@@ -267,6 +267,60 @@ int bus_connect_user_systemd(sd_bus **ret_bus) {
         return 0;
 }
 
+int bus_connect_project_systemd(const char *project, sd_bus **ret_bus) {
+        _cleanup_(sd_bus_close_unrefp) sd_bus *bus = NULL;
+        _cleanup_free_ char *pp = NULL;
+        int r;
+
```
Comment (bluca) [inline src/shared/bus-util.c:313]: assert(ret_bus)

**Comment by bluca** [comment]
Given it's intended for non-interactive users, maybe `headless-session@.service` or `headless@.service`? That's how non-interactive systems are usually called, so it would be immediately clear what it is for

**Comment by DaanDeMeyer** [comment]
> Given it's intended for non-interactive users, maybe headless-session@.service or headless@.service? That's how non-interactive systems are usually called, so it would be immediately clear what it is for

What do you name the CLI switch then though?

**Comment by bluca** [comment]
> > Given it's intended for non-interactive users, maybe [headless-session@.service](mailto:headless-session@.service) or [headless@.service](mailto:headless@.service)? That's how non-interactive systems are usually called, so it would be immediately clear what it is for
> 
> What do you name the CLI switch then though?

`systemctl start --headless foo.service` seems nice and intuitive to me

**Comment by poettering** [comment]
> Given it's intended for non-interactive users, maybe `headless-session@.service` or `headless@.service`? That's how non-interactive systems are usually called, so it would be immediately clear what it is for

But user@.service is also headless...

I think it's essential to convey that this isn't for regular users (which are just run headless) but implies a new thing, i.e. a project uid, and a project home dir and so on.

**Comment by poettering** [comment]
> `systemctl start --headless foo.service` seems nice and intuitive to me

i don't understand what that would do?

**Comment by poettering** [comment]
btw, i stole the "project" name from the "project quota" thing that solaris (and now linux has)

**Comment by bluca** [comment]
> > Given it's intended for non-interactive users, maybe `headless-session@.service` or `headless@.service`? That's how non-interactive systems are usually called, so it would be immediately clear what it is for
> 
> But [user@.service](mailto:user@.service) is also headless...
> 
> I think it's essential to convey that this isn't for regular users (which are just run headless) but implies a new thing, i.e. a project uid, and a project home dir and so on.

The regular one uses logind and pam, no? Isn't that the key difference?

**Comment by poettering** [comment]
> > 
> 
> The regular one uses logind and pam, no? Isn't that the key difference?

That's one major difference. But what is super important is that it uses DynamicUser=1, i.e. you do not invoke it for an existing user such as user@.service, but you invoke it jsut like that, and the user then gets transienlty created and goes away afterwards. And I think that's at least as important: its a new world we live in.

**Comment by DaanDeMeyer** [comment]
> That's one major difference. But what is super important is that it uses DynamicUser=1, i.e. you do not invoke it for an existing user such as [user@.service](mailto:user@.service), but you invoke it jsut like that, and the user then gets transienlty created and goes away afterwards. And I think that's at least as important: its a new world we live in.

Is this a requirement? It'd be great if we could do both. Either you instantiate this for a user that already exists and then we run as that user, or you instantiate this for a user that does not exist and in that case we do the dynamic user logic

**Comment by poettering** [comment]
I force pushed a new version of this. This should be ready for review now.

v256 material.

I quite like it. It's really simple, but feels very powerful. I made systemctl, systemd-run and busctl all work with projects, you now have a --project= switch on each to connect to some project, and do stuff there.

I added full docs.

CI test stil missing.

I answered some of my earlier questions:

> * Do we actually want this? It might confuse people, since they now have have another way to bundle up things:

yes, I think we do. It's simple and quite nice to work with.

> * dbus config prohibits access to all foreign users, including those with UID 0. Which means you cannot actually talk to the dbus instance we start for the project from the outside. I haven't found a nice way to relax this policy for project busses without also relaxing it for [user@.service](mailto:user@.service), i.e. regular users.

I solved this: connections to relevant D-Bus sockets are now done under the UID of the project. Hence everything just works now.

> * I am not even sure about the name of the concept, is "project" really that great?

I couldn't find something better, and I think the "project quota" concept kinda made sense to use the word "project", so I think we can reuse this here.

> * Which slice should projects live in by default? in system.slice? in user.slice? in a new top-level slice? For now I picked system-projects.slice, but I am not sure this is the best approach.

I moved this to project.slice now.

> * Should we maybe default to start something other than `default.target` in the service manager? maybe `project@%i.target`? After all, this service manager is mostly just another --user instance, hence if people add services to system-wide per-user default.target they might be quite confused what they get.

Yes, I implemented this. Project service managers now invoke the project@%i.target user unit by default.

> * Should the service maybe be called `project@.service` rather than `systemd-project@.service`? The rules around naming services are a bit blurry, but the idea was always that generic concepts should not be prefixed with `systemd-`, and systemd specific implementations should be. From logind's perspective [user@.service](mailto:user@.service) was generic and hence has no prefix. Question is what to do about the project service manager? with or without prefix?

I dropped the prefix, the project service manager is now called project@.service.

Anyway, please have a look, check out the docs. It's a fun tool

**Comment by yuwata** [comment]
packit failures:
```
1256/1260 systemd:dist / check-help-systemd-vmspawn                   FAIL              0.02s   exit status 1
>>> LD_LIBRARY_PATH=/builddir/build/BUILD/systemd-255.rc1/redhat-linux-build/src/shared MALLOC_PERTURB_=157 /builddir/build/BUILD/systemd-255.rc1/tools/check-help.sh /builddir/build/BUILD/systemd-255.rc1/redhat-linux-build/systemd-vmspawn
systemd-vmspawn --help output is too wide:
     --secure-boot=         Configure whether to search for firmware which supports Secure Boot

```
CentOS CI failure is #28472, unrelated.

**Comment by YHNdnzj** [comment]
> packit failures:
> ```
> 1256/1260 systemd:dist / check-help-systemd-vmspawn                   FAIL              0.02s   exit status 1
> >>> LD_LIBRARY_PATH=/builddir/build/BUILD/systemd-255.rc1/redhat-linux-build/src/shared MALLOC_PERTURB_=157 /builddir/build/BUILD/systemd-255.rc1/tools/check-help.sh /builddir/build/BUILD/systemd-255.rc1/redhat-linux-build/systemd-vmspawn
> systemd-vmspawn --help output is too wide:
>      --secure-boot=         Configure whether to search for firmware which supports Secure Boot
> 
> ```

Fixed by https://github.com/systemd/systemd/commit/6527d175cda8c2e1feceb26eb7e3ec111ddc6ae4. Just needs a rebase.

**Comment by poettering** [comment]
> CI test still missing.

Added now too.

**Comment by bluca** [comment]
The name as is will make it all but impossible to search for this - googling "systemd project documentation" or so will never return the right thing. I find it confusing and ambiguous. Given this is the counter part of the user session, I'd call if something like: headless, headless session, anonymous session, autonomous session, auto session, detached session

**Comment by poettering** [comment]
> The name as is will make it all but impossible to search for this - googling "systemd project documentation" or so will never return the right thing. I find it confusing and ambiguous. Given this is the counter part of the user session, I'd call if something like: headless, headless session, anonymous session, autonomous session, auto session, detached session

session is definitely the wrong word for this. There are multiple definitions of "session" in our world already, logind sessions, utmp sessions, audit sessions, pam sessions, setsid sessions. They are all a bit different but somewhat overlap. But this stuff here doesn't overlap at all, by none of these definitions this is a session. The stuff does not register as a session with utmp, logind, audit, pam nor setsid.

Hence, anything with "session" in it would be an absolute misnomer, and be against any of the concepts we named that way so far.

And again, headless sounds wrong too, since it's really not that. in fact, it could even have a "head" if you want. Any regular user sessions could also be headless.

sorry, but "headless" and "session" are both things that are very misleading, and worse than "project"

**Comment by bluca** [comment]
What about 'instance' instead of 'session' then - autonomous instance, anonymous instance, headless instance, etc, that should be googleable

**Comment by qhill** [comment]
dynamic-user@.service?

**Comment by bluca** [comment]
> [dynamic-user@.service](mailto:dynamic-user@.service)?

that's not bad at all - unique, searchable, self-descriptive, short and to the point

**Comment by poettering** [comment]
> [dynamic-user@.service](mailto:dynamic-user@.service)?

This suggests that this was used to implement DynamicUser=1, but it's really not. This is one *use* of DynamicUser=1, but just one of many. It would be like naming the concept of a knife potato-cutter, just because it can cut potatoes, but it certainly can do a lot more.

**Comment by bluca** [comment]
user@.service doesn't imply that it is used to implement User=, so I don't think that is any different

File: man/project@.service.xml [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+  <refsynopsisdiv>
+    <para><filename>project@<replaceable>NAME</replaceable>.service</filename></para>
+  </refsynopsisdiv>
+
+  <refsect1>
+    <title>Description</title>
+
+    <para>Service managers for projects run in
+    <filename>project@<replaceable>NAME</replaceable>.service</filename> system units, with the project name as the
+    instance identifier. Projects are way to run additional instances of the service manager, under dynamic
```
Comment (behrmann) [inline man/project@.service.xml]:
```suggestion
    instance identifier. Projects are a way to run additional instances of the service manager, under dynamic
```

File: man/project@.service.xml [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    for example with the <option>--what=all</option> switch.</para>
+
+    <para>The <filename>project@.service</filename> unit invokes a <command>systemd --user</command>
+    service manager process. This means unit files are looked for according to the sames rules as for regular user
+    service managers, for example in
+    <filename>/var/lib/projects/<replaceable>NAME</replaceable>/.config/systemd/user/</filename>.</para>
+
+    <para>Project names may be chosen freely by the user, however, they must be suitable as UNIX filenames
+    (i.e. 255 characters max, and contain no <literal>/</literal>), and when prefixed with
+    <literal>p-</literal> be suitable as a user name matching strict POSIX rules, see <ulink
```
Comment (behrmann) [inline man/project@.service.xml]: I guess that should be 
```suggestion
    <literal>p_</literal> be suitable as a user name matching strict POSIX rules, see <ulink
```

**Comment by TriMoon** [comment]
@poettering 
> ideas for all these questions more than welcome!

Reading the discussion so far, and the intended use-cases:
- Wouldn't it be better to name the feature/service template `hardened` instead?
That will even allow you to make other defaults for it besides `DynamicUser=1` :wink::thinking:

**Comment by Winterhuman** [comment]
What about `user-less@.service`? It wouldn't strictly be tied to `DynamicUser=1`, but it should still convey the ephemeral identity feature

**Comment by TriMoon** [comment]
> It's not "hardened" at all by any definiton of the word I know. I mean, sure it uses DynamicUser=1, but it uses _none_ of our namespacing options, and I don#t think it should.
> 
> It's about starting a new service manager, under its own ephemeral identity, that's all.

Then I and maybe others don't actually understand what the end-goal purpose of your idea is exactly...

Isn't your goal to provide a uniform way for daemon-service writers to make use of "User" restrictions, which is mostly used to harden their daemon's? :thinking: 
Sure they already can and do now, except for the random user part, and as far as i understand your idea is to give them a "standard" way in a new service manager...

**Comment by Winterhuman** [comment]
From what I understand from the description item: "It has HOME set to /var/lib/projects/%i and XDG_RUNTIME_DIR set to /run/projects/%i".

`project@.service` is, for the most part, just about creating non-root services which aren't stored under `/home/some-user/.config/systemd/`, and therefore, they aren't tied to a specific login user. The service gets to be non-root *and* have system-wide reach at the same time.

As an (admittedly niche) example, a service running `makepkg` has to be non-root since that command can't be run as root, but, it would be useful for the whole system and not any one user.

**Comment by Werkov** [comment]
subsystemd@.service :bike:

**Comment by mrc0mmand** [comment]
Given we can't decide on a name, we should maybe switch to emojis and call it something like `🚀@.service` (I'm surprised Lennart hasn't proposed that already).

**Comment by poettering** [comment]
Maybe we should call it pod? taking a page out of the docker/podman book?

**Comment by bluca** [comment]
That's not bad (although nowhere near as good as 🚀@.service), but have you checked that podman doesn't already ship something in that namespace?

**Comment by mrc0mmand** [comment]
Maybe something like workspace or playground would work better, so we avoid the association with podman/kubernetes?

**Comment by bluca** [comment]
WORKGROUPS! SYSTEMD FOR WORKGROUPS! DO IT!

**Comment by poettering** [comment]
Capsule?

**Comment by poettering** [comment]
Aggregate?

**Comment by poettering** [comment]
Compound?

**Comment by poettering** [comment]
Composite?

**Comment by poettering** [comment]
Batch? Bundle? pack?

**Comment by poettering** [comment]
Hoard? Trove? Inventory?

**Comment by bluca** [comment]
WORKGROUPS

systemd_for_workgroups

**Comment by poettering** [comment]
Zone?

**Comment by poettering** [comment]
Realm?

**Comment by bluca** [comment]
I made you the logo and cover, so you can ship floppy disk boxes

**Comment by dtardon** [comment]
Compartment? Cell? Domain?

From the already proposed names, I like capsule the most (even though it doesn't come with a ready-made logo ;-)

**Comment by behrmann** [comment]
I think I could live with most of the proposed names but I would caution against pod. I think this would lead to a lot of confusion among people potentially using this. For a similar reason, but to a lesser extent, I would be against bundle, since the package manager/virtual environment tool in Ruby is called bundler.

**Comment by bluca** [comment]
> I think I could live with most of the proposed names but I would caution against pod. I think this would lead to a lot of confusion among people potentially using this. For a similar reason, but to a lesser extent, I would be against bundle, since the package manager/virtual environment tool in Ruby is called bundler.

another reason why WORKGROUPS is the only obvious and reasonable choice

**Comment by arianvp** [comment]
I'd love if the user could pick the directory where project picks up unit files. E.g. Say I have source code for my service `~/Projects/my-super-cool-app` that I can have my unit files in `~/Projects/my-super-cool/app/units/my-cool-app.service` and run the "project" from my development folder. e.g. to spawn up my app and a postgres database whilst testing.

**Comment by TriMoon** [comment]
> ay I have source code for my service `~/Projects/my-super-cool-app` that I can have my unit files in `~/Projects/my-super-cool-app/.config/systemd/user/my-cool-app.service` and run the "project" from my development folder. e.g. to spawn up my app and a postgres database whilst testing.
> 
> This also allows me to check these units in into my git repo to share with my fellow engineers.

You can already do that all, you just need to "bind-mount" your directory onto the correct spot in the system while you're working on them...

**Comment by claes** [comment]
When something descriptive seems too generic or not accurate enough, I think it may be good to coin a new word for it. If it is about users, but not of flesh and blood, I am thinking of ghosts, angels, spirits. When looking for words similar to daemon, I found the word numen, which I didn't know about:  

""Numen," in ancient Roman religion, refers to a divine presence or the spirit of a deity residing in a particular object or place. The concept is less about a distinct anthropomorphic god and more about the inherent divine power or force that can exist in anything. It represents the idea that everything in nature, from trees and rivers to objects and places, can have its own divine essence or spiritual force."

Not sure if a word like that exactly describes what this is about, but it is distinct enough that it will not be confused with the systemd Github project.

**Comment by poettering** [comment]
I am kinda leaning towards "zone" btw. Is there major opposition to that?

**Comment by bluca** [comment]
> WORKGROUPS
> 
> systemd_for_workgroups

Don't make me tap the sign

**Comment by YHNdnzj** [comment]
I like capsule and workgroup a lot and can't decide between them :)

**Comment by Werkov** [comment]
FTR, the term [zone](https://docs.oracle.com/cd/E18440_01/doc.111/e18415/chapter_zones.htm#OPCUG426) is used for a thing that resembles containers (with all namespaces and stuff) more than a separate systemd instance.

**Comment by poettering** [comment]
renamed the feature "capsule" now. seems to me there was the least opposition to. ptal.

**Comment by bluca** [comment]
...I even drew the logo!

**Comment by poettering** [comment]
don't make me change this again, one `sed -i` orgy like this was enough for this week.

**Comment by mrc0mmand** [comment]
@poettering this should address the ASan fail:

```diff
diff --git a/test/units/testsuite-74.capsule.sh b/test/units/testsuite-74.capsule.sh
index d3c6bee89a..95ca3d08d5 100755
--- a/test/units/testsuite-74.capsule.sh
+++ b/test/units/testsuite-74.capsule.sh
@@ -13,6 +13,14 @@ at_exit() {
 
 trap at_exit EXIT
 
+# Appease ASan, since the capsule@.service uses DynamicUser=yes
+mkdir -p /run/systemd/system/capsule@.service.d/
+cat >/run/systemd/system/capsule@.service.d/99-asan.conf <<EOF
+[Service]
+EnvironmentFile=-/usr/lib/systemd/systemd-asan-env
+EOF
+systemctl daemon-reload
+
 (! test -f /run/capsules/foobar )
 (! test -f /var/lib/capsules/foobar )
 (! id -u c-foobar )
```

**Comment by YHNdnzj** [comment]
@mrc0mmand This can now be shortened to the following:

```bash
systemctl edit --stdin capsule@.service --drop-in=99-asan.conf <<EOF
[Service]
EnvironmentFile=-/usr/lib/systemd/systemd-asan-env
EOF
```
 ;-)

**Comment by mrc0mmand** [comment]
> @mrc0mmand This can now be shortened to the following:
> 
> ```shell
> systemctl edit --stdin capsule@.service --drop-in=99-asan.conf <<EOF
> [Service]
> EnvironmentFile=-/usr/lib/systemd/systemd-asan-env
> EOF
> ```
> 
> Without the need to mkdir or reload explicitly ;-)

Hah, that's nifty, will need to remember that!

**Comment by poettering** [comment]
Thanks so much, @mrc0mmand, @YHNdnzj!

**Comment by amluto** [comment]
I'm contemplating using this and/or portable services for system I operate, and I'm struggling a bit.  I think the issue is that there are a bunch of things going on that seem like they aren't quite orthogonal enough.  Here are the features I think I'd like, but I'm sort of cherry-picking:

1. Building the image.  Right now I'm using podman.  I really wish everyone could agree on conventions as to what actually belongs in the image.  For example, podman seems to create /etc/hostname by default, and I have to actively fight it to get rid of it.  nspawn does _not_ want /etc/hostname.  podman and portable services don't need a DHCP client inside.  nspawn does (and I keep meaning to file an RFE to get rid of that requirement).  But whatever -- I have this working.

2. Arranging to deploy and update my image.  I can use machinectl to unpack into /var/lib/machines or I can use portable services to "attach" and partially unpack.  I can manually start and stop things and fiddle with symlinks for updating.  It's not great but it works.  (Podman, Docker, etc handle this sort of nicely but are really nasty in some of the details.)

3. Introspecting the image, debugging it, etc.  Capsule is neutral.  Portable services seem quite obnoxious -- see below.  nspawn works great. 

4. Running the thing.  nspawn gets everything right.  Capsules seem like they would too.  Portable services could, I suppose, contain a capsule.  Really, a capsule seems a bit like a systemd feature that lets me bundle a bunch of stuff up into a systemd instance, and IMO it would be nice for that to work cleanly and easily inside a portable service or a machine.

--- Introspection (referenced above) ---

The thing I actually want to encapsulate (sorry) is the fs namespace.  Linux supports fs namespaces as sort-of first class objects.  So IMO I _should_ be able to bundle my software up as an image or directory, have systemd (or whatever) create the fs namespace, set up the mounts that need setting up from the outside, set up whatever overlays are needed, etc, and keep the namespace around with a name (e.g. bind-mounting the namespace somewhere useful such that nsenter, for example, could use it).  And _not_ run anything from inside the image.

Then nspawn could boot the thing, but I could also nsenter it without booting it.  And portable services could invoke things inside, but without all the complication of RootImage=, but instead (presumably with a degree of assistance from systemd knowing how things are organized) using a new option a la FSNamespace= to enter the correct namespace.  And capsules could point at it via FSNamespace=.



So I suppose what I'm suggesting in general is that I think this would all be more useful if the actual functionality of all of systemd's various grouping and image handling mechanisms were broken down into orthogonal-ish features that could be composed to build what needs building.

**Comment by bluca** [comment]
@amluto

> Building the image

Our solution for this is mkosi https://github.com/systemd/mkosi
In general, anything that can build DDIs will do https://uapi-group.org/specifications/specs/discoverable_disk_image/

> Arranging to deploy and update my image.

Depends on how you want to instantiate them. We have machined for nspawn/vmspawn for fully isolated payloads, and portabled for fully integrated payloads.
The downloading story is weaker, but there is something rudimentary available with importd and sysupdated.

> Introspecting the image, debugging it

These are GPT images, so anything that can open them should provide everything you need? And there's various ways to just start a shell on a DDI, both nspawn --image and systemd-run --shell -p RootImage allow to do just that.

> Running the thing

As above, it depends on _how_ you want to run them. Fully integrated? Portabled. Same kernel but otherwise segregated? Nspawn. Fully segregated? Vmspawn.

> So I suppose what I'm suggesting in general is that I think this would all be more useful if the actual functionality of all of systemd's various grouping and image handling mechanisms were broken down into orthogonal-ish features that could be composed to build what needs building.

That means you leak things left and right. In general, the runtime resources are all associated with the execution payload, and get automatically cleaned up when the execution payload goes away. That is a very fundamental and good property to have. In some previous iterations of stackable portables I had exactly that, with a bunch of stuff being mounted to form full filesystems in /run/ and then later running things inside of them. It was an unmitigated disaster, with random programs (and users!) taking random dependencies and keeping things pinned, making it impossible to account for usage, do upgrades, cleanly shutdown, and whatnot. It's definitely a bad pattern and it was a huge mistake to set it up that way.

The key concept with what we offer here instead is that it's trivially easy (and trivially cheap, despite dm-verity and whatnot) to get an _equivalent_ environment set up and running, as many times as you want, at the same time or not. With full resource accounting and automated cleanups.

**Comment by amluto** [comment]
@bluca

> > Introspecting the image, debugging it
> 
> These are GPT images, so anything that can open them should provide everything you need? And there's various ways to just start a shell on a DDI, both nspawn --image and systemd-run --shell -p RootImage allow to do just that.

Sorry, I wasn't clear.  I don't mean introspecting an image per se -- I meant introspecting the _running_ container/capsule/machine/portable service.  There's machinectl --shell, which is fairly weak if the tooling I want isn't installed in the image.  (To be fair, systemd's competitors are weak here, too.)

> 
> > Running the thing
> 
> As above, it depends on _how_ you want to run them. Fully integrated? Portabled. Same kernel but otherwise segregated? Nspawn. Fully segregated? Vmspawn.
> 
> > So I suppose what I'm suggesting in general is that I think this would all be more useful if the actual functionality of all of systemd's various grouping and image handling mechanisms were broken down into orthogonal-ish features that could be composed to build what needs building.
> 
> That means you leak things left and right. In general, the runtime resources are all associated with the execution payload, and get automatically cleaned up when the execution payload goes away. That is a very fundamental and good property to have. In some previous iterations of stackable portables I had exactly that, with a bunch of stuff being mounted to form full filesystems in /run/ and then later running things inside of them. It was an unmitigated disaster, with random programs (and users!) taking random dependencies and keeping things pinned, making it impossible to account for usage, do upgrades, cleanly shutdown, and whatnot. It's definitely a bad pattern and it was a huge mistake to set it up that way.

Oh, I fully agree that, in the old way of doing things, this would be a huge mess.

But it's 2024, and this problem has a nice solution: _we have systemd!_.   A a sketch of how this could work, imagine that a set-up-and-mounted group of namespaces is called a namespace-pile.  (Please don't name it that if this ever lands in systemd for real!)

systemd-namespace-pile@mypackage.namespace-pile: This is a unit that represents a userns (if private userns is enabled), a mount ns, etc.

myservice.service: A service running from those namespaces, perhaps lile:

```
Namespaces=systemd-namespace-pile@mypackage.namespace-pile
```

And systemd arranges for this to be a real dependency, and shutting down the namespace-pile actually cleans everything up.  And I can have more than one service in the same namespace-pile if I really want -- it's up to me.

Debugging can use nsenter or systemd-aware variants of it that work on namespace-piles.  It would be straightforward to write new tools that would do fancy things like opening a shell where the namespace-pile is mounted (as a recursive mount) somewhere like /var/tmp/my-view-into-the-pile

> 
> The key concept with what we offer here instead is that it's trivially easy (and trivially cheap, despite dm-verity and whatnot) to get an _equivalent_ environment set up and running, as many times as you want, at the same time or not. With full resource accounting and automated cleanups.

But I don't just want _equivalent_ environments set up!  I want multiple invocations of the _same_ instance of the environment set up, so I can have multiple running at once, cooperatively, in the environment.

Really I want a two-level hierarchy, where there are instantiations of the environment, and then there are services (and shells, etc) running attached to that instantiation of the environment.

As an example of my current production usecase, I have an image (the one I'm building with podman, but it could probably be mkosi if I were so inclined).  In that image, I have exactly zero things that need to run as root (or as "root" in a namespace). I have quite a few service-like things that need to run from the instantiation of the image, and they need to run from the _same_ instantiation of the image, because they cooperate, and they will flat-out not work if they are run from separate instantiations [0].  And I don't want a netns (which is mostly irrelevant to this discussion, but see below).

So I'm currently using systemd-nspawn and I'm running my own service manager (which predates systemd, and works quite well for my use case) inside the machine.  And everything running inside the machine as root is pure overhead and annoyance (especially since I'm not using a netns).  (Except that I'm using the PID 1 systemd instance of systemd inside the container to run my service manager as a service -- the 'user' systemd won't land me in the right place in the cgroup hierarchy -- maybe capsules would be better here.)

But it would all start faster and be more convenient if this were just a 'namespace-pile' and I could skip all the parts of nspawn that I don't actually need for my use case.


[0] I could probably re-architect this, but I don't really want to, and the whole thing would have more overhead if I rearchitected this.

**Comment by bluca** [comment]
> And systemd arranges for this to be a real dependency, and shutting down the namespace-pile actually cleans everything up. And I can have more than one service in the same namespace-pile if I really want -- it's up to me.

Been there, done that - sounds good, doesn't work. It always ends in a stupid mess, and it's not worth it. As you said yourself, it's just that you got some old usecase that you don't want to update. IE: it's the inertia of a legacy setup driving that.

If you want to run multiple services in the same image, use nspawn - that's what it's for. You can add as many services as you want to the image, and they'll all happily run together.

**Review by keszybz — APPROVED** [review]
LGTM.

The man page text could use a little polish, but it's nothing major to hold up the pull request over.

I also am not convinced that a new top-level slice should be used. If those things are going to be used for system services, then I think they should go under `/system/capsule`. Or maybe there even shouldn't be any separate capsule slice, and `capsule@.service` should be directly under `system.slice`. Essentially, I think we should group by their purpose, not by type. But this is something that we can change later once this starts being used and we have more idea what it is being used *for*.

There was a lot of bikeshedding over the name… I think capsule is not a bad name. It's certainly unique and googleable.

I think we should try this, and if it turns out that it doesn't work as expected, we can always remove it or rework.

**Pull request merged**